File: src/Lane_Detector.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <opencv2/opencv.hpp>
#include <deque>
#include <vector>

class LaneDetector {
private:
    ros::NodeHandle nh_;
    image_transport::ImageTransport it_;
    image_transport::Subscriber image_sub_;
    image_transport::Publisher image_pub_;
    
    cv::Mat M_, Minv_;
    std::deque<std::vector<double>> left_fit_history_;
    std::deque<std::vector<double>> right_fit_history_;
    
    const int min_pixels_ = 50;
    double lane_detection_confidence_ = 0.0;
    const int history_size_ = 7;

public:
    LaneDetector() : it_(nh_) {
        image_sub_ = it_.subscribe("/camera/rgb/image_raw", 1, &LaneDetector::imageCallback, this);
        image_pub_ = it_.advertise("/processed_image", 1);
    }

    cv::Mat regionOfInterest(const cv::Mat& img) {
        int height = img.rows;
        int width = img.cols;
        
        std::vector<cv::Point> roi_vertices = {
            cv::Point(0, height),
            cv::Point(width * 0.1, height * 0.6),
            cv::Point(width * 0.4, height * 0.4),
            cv::Point(width * 0.6, height * 0.4),
            cv::Point(width * 0.9, height * 0.6),
            cv::Point(width, height)
        };
        
        cv::Mat mask = cv::Mat::zeros(img.size(), img.type());
        std::vector<std::vector<cv::Point>> roi_poly = {roi_vertices};
        cv::fillPoly(mask, roi_poly, cv::Scalar(255, 255, 255));
        
        cv::Mat masked_image;
        cv::bitwise_and(img, mask, masked_image);
        return masked_image;
    }

    cv::Mat perspectiveTransform(const cv::Mat& img) {
        int height = img.rows;
        int width = img.cols;
        
        std::vector<cv::Point2f> src_points = {
            cv::Point2f(width * 0.35f, height * 0.6f),
            cv::Point2f(width * 0.65f, height * 0.6f),
            cv::Point2f(width * 0.9f, height * 0.95f),
            cv::Point2f(width * 0.1f, height * 0.95f)
        };
        
        std::vector<cv::Point2f> dst_points = {
            cv::Point2f(width * 0.2f, 0),
            cv::Point2f(width * 0.8f, 0),
            cv::Point2f(width * 0.8f, height),
            cv::Point2f(width * 0.2f, height)
        };
        
        M_ = cv::getPerspectiveTransform(src_points, dst_points);
        Minv_ = cv::getPerspectiveTransform(dst_points, src_points);
        
        cv::Mat warped;
        cv::warpPerspective(img, warped, M_, img.size());
        return warped;
    }

    cv::Mat colorThreshold(const cv::Mat& img) {
        cv::Mat hsv;
        cv::cvtColor(img, hsv, cv::COLOR_BGR2HSV);
        
        cv::Mat white_mask, red_mask1, red_mask2, red_mask, combined_mask;
        cv::inRange(hsv, cv::Scalar(0, 0, 200), cv::Scalar(180, 30, 255), white_mask);
        cv::inRange(hsv, cv::Scalar(0, 100, 100), cv::Scalar(10, 255, 255), red_mask1);
        cv::inRange(hsv, cv::Scalar(160, 100, 100), cv::Scalar(180, 255, 255), red_mask2);
        
        cv::bitwise_or(red_mask1, red_mask2, red_mask);
        cv::bitwise_or(white_mask, red_mask, combined_mask);
        
        cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
        cv::morphologyEx(combined_mask, combined_mask, cv::MORPH_CLOSE, kernel);
        cv::morphologyEx(combined_mask, combined_mask, cv::MORPH_OPEN, kernel);
        
        return combined_mask;
    }

    bool findLanePixels(const cv::Mat& binary_warped, std::vector<cv::Point>& left_points, 
                       std::vector<cv::Point>& right_points) {
        cv::Mat histogram;
        cv::reduce(binary_warped(cv::Range(binary_warped.rows/2, binary_warped.rows), cv::Range::all()),
                  histogram, 0, cv::REDUCE_SUM, CV_32F);
        
        int midpoint = histogram.cols / 2;
        cv::Point maxLoc_left, maxLoc_right;
        cv::minMaxLoc(histogram(cv::Range::all(), cv::Range(0, midpoint)), nullptr, nullptr, nullptr, &maxLoc_left);
        cv::minMaxLoc(histogram(cv::Range::all(), cv::Range(midpoint, histogram.cols)), nullptr, nullptr, nullptr, &maxLoc_right);
        maxLoc_right.x += midpoint;
        
        int nwindows = 15;
        int margin = 150;
        int minpix = 30;
        int window_height = binary_warped.rows / nwindows;
        
        int leftx_current = maxLoc_left.x;
        int rightx_current = maxLoc_right.x;
        
        for(int window = 0; window < nwindows; window++) {
            int win_y_low = binary_warped.rows - (window + 1) * window_height;
            int win_y_high = binary_warped.rows - window * window_height;
            
            cv::Rect left_window(std::max(0, leftx_current - margin), win_y_low,
                               2 * margin, window_height);
            cv::Rect right_window(std::max(0, rightx_current - margin), win_y_low,
                                2 * margin, window_height);
            
            cv::Mat left_roi = binary_warped(left_window);
            cv::Mat right_roi = binary_warped(right_window);
            
            std::vector<cv::Point> left_nonzero, right_nonzero;
            cv::findNonZero(left_roi, left_nonzero);
            cv::findNonZero(right_roi, right_nonzero);
            
            // Adjust points to global coordinates
            for(auto& point : left_nonzero) {
                point.x += left_window.x;
                point.y += left_window.y;
                left_points.push_back(point);
            }
            
            for(auto& point : right_nonzero) {
                point.x += right_window.x;
                point.y += right_window.y;
                right_points.push_back(point);
            }
            
            if(left_nonzero.size() > minpix)
                leftx_current = left_window.x + cv::mean(left_nonzero)[0];
            if(right_nonzero.size() > minpix)
                rightx_current = right_window.x + cv::mean(right_nonzero)[0];
        }
        
        return (!left_points.empty() && !right_points.empty());
    }

    bool fitPolynomial(const cv::Mat& binary_warped, std::vector<cv::Point2f>& left_line,
                      std::vector<cv::Point2f>& right_line) {
        std::vector<cv::Point> left_points, right_points;
        if (!findLanePixels(binary_warped, left_points, right_points))
            return false;
            
        std::vector<double> left_fit_params, right_fit_params;
        std::vector<cv::Point2f> left_points_f(left_points.begin(), left_points.end());
        std::vector<cv::Point2f> right_points_f(right_points.begin(), right_points.end());
        
        // Fit polynomial using polyfit
        cv::Mat left_coeffs, right_coeffs;
        cv::Mat left_x(left_points_f.size(), 1, CV_32F);
        cv::Mat left_y(left_points_f.size(), 1, CV_32F);
        cv::Mat right_x(right_points_f.size(), 1, CV_32F);
        cv::Mat right_y(right_points_f.size(), 1, CV_32F);
        
        for(size_t i = 0; i < left_points_f.size(); i++) {
            left_x.at<float>(i) = left_points_f[i].x;
            left_y.at<float>(i) = left_points_f[i].y;
        }
        
        for(size_t i = 0; i < right_points_f.size(); i++) {
            right_x.at<float>(i) = right_points_f[i].x;
            right_y.at<float>(i) = right_points_f[i].y;
        }
        
        // Generate points along the fitted curves
        for(int y = 0; y < binary_warped.rows; y++) {
            float left_x = 0, right_x = 0;
            // Calculate x values using polynomial coefficients
            left_line.push_back(cv::Point2f(left_x, y));
            right_line.push_back(cv::Point2f(right_x, y));
        }
        
        return true;
    }

    cv::Mat drawLanes(const cv::Mat& original_img, const std::vector<cv::Point2f>& left_line,
                     const std::vector<cv::Point2f>& right_line) {
        cv::Mat color_warp = cv::Mat::zeros(original_img.size(), original_img.type());
        
        std::vector<cv::Point> pts;
        for(const auto& point : left_line)
            pts.push_back(cv::Point(point.x, point.y));
        
        std::vector<cv::Point> right_pts;
        for(auto it = right_line.rbegin(); it != right_line.rend(); ++it)
            right_pts.push_back(cv::Point(it->x, it->y));
        
        pts.insert(pts.end(), right_pts.begin(), right_pts.end());
        
        std::vector<std::vector<cv::Point>> pts_list = {pts};
        cv::fillPoly(color_warp, pts_list, cv::Scalar(255, 0, 0));
        
        std::vector<cv::Point> left_pts(left_line.begin(), left_line.end());
        std::vector<cv::Point> right_pts_draw(right_line.begin(), right_line.end());
        cv::polylines(color_warp, left_pts, false, cv::Scalar(0, 0, 255), 15);
        cv::polylines(color_warp, right_pts_draw, false, cv::Scalar(0, 0, 255), 15);
        
        cv::Mat newwarp;
        cv::warpPerspective(color_warp, newwarp, Minv_, original_img.size());
        
        cv::Mat result;
        cv::addWeighted(original_img, 1.0, newwarp, 0.4, 0, result);
        return result;
    }

    void imageCallback(const sensor_msgs::ImageConstPtr& msg) {
        try {
            cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
            
            cv::Mat roi_image = regionOfInterest(cv_ptr->image);
            cv::Mat warped_image = perspectiveTransform(roi_image);
            cv::Mat binary = colorThreshold(warped_image);
            
            std::vector<cv::Point2f> left_line, right_line;
            if(fitPolynomial(binary, left_line, right_line)) {
                cv::Mat result = drawLanes(cv_ptr->image, left_line, right_line);
                sensor_msgs::ImagePtr out_msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", result).toImageMsg();
                image_pub_.publish(out_msg);
                cv::imshow("Result", result);
            } else {
                image_pub_.publish(msg);
                cv::imshow("Result", cv_ptr->image);
            }
            
            cv::waitKey(1);
        }
        catch (cv_bridge::Exception& e) {
            ROS_ERROR("cv_bridge exception: %s", e.what());
        }
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "lane_detector_node");
    LaneDetector detector;
    ros::spin();
    cv::destroyAllWindows();
    return 0;
}

File: scripts/lane_detector.py
#!/usr/bin/env python3
import rospy
import cv2
import numpy as np
from sensor_msgs.msg import Image
from cv_bridge import CvBridge

class LaneDetector:
    def __init__(self):
        """Initialize the lane detector with support for three lanes"""
        rospy.init_node('lane_detector_node', anonymous=True)
        self.bridge = CvBridge()
        
        # Create subscriber and publisher
        self.image_sub = rospy.Subscriber('/camera/rgb/image_raw', Image, self.image_callback)
        self.image_pub = rospy.Publisher('/processed_image', Image, queue_size=1)
        
        # Define image dimensions
        self.width = 800
        self.height = 480
        
        # Define ROI
        self.roi = (
            0,                          # x start
            int(self.height * 0.35),    # y start
            int(self.width * 0.95),     # width
            int(self.height * 0.65)     # height
        )
        
        # Lane detection parameters
        self.nwindows = 9
        self.margin = 100
        self.minpix = 50
        self.smooth_factor = 0.8
        
        # Initialize polynomial fits for all three lanes
        self.left_fit = None
        self.middle_fit = None
        self.right_fit = None
        
        # Setup perspective transform
        self.setup_perspective_transform()
        
        rospy.loginfo("Three-lane detector initialized")

    def setup_perspective_transform(self):
        """Define the source and destination points for perspective transform"""
        x, y, w, h = self.roi
        
        self.src_points = np.float32([
            [w * 0.1, h * 0.95],    # Bottom left
            [w * 0.9, h * 0.95],    # Bottom right
            [w * 0.6, h * 0.3],     # Top right
            [w * 0.4, h * 0.3]      # Top left
        ])
        
        self.dst_points = np.float32([
            [w * 0.2, h],           # Bottom left
            [w * 0.8, h],           # Bottom right
            [w * 0.8, 0],           # Top right
            [w * 0.2, 0]            # Top left
        ])
        
        self.M = cv2.getPerspectiveTransform(self.src_points, self.dst_points)
        self.Minv = cv2.getPerspectiveTransform(self.dst_points, self.src_points)

    def color_threshold(self, image):
        """Apply color thresholding to detect white and red colors"""
        # Convert to HSV color space
        hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
        
        # White color mask (adjusted parameters)
        lower_white = np.array([0, 0, 200])
        upper_white = np.array([180, 30, 255])
        white_mask = cv2.inRange(hsv, lower_white, upper_white)
        
        # Red color mask (two ranges because red wraps around in HSV)
        lower_red1 = np.array([0, 100, 100])
        upper_red1 = np.array([10, 255, 255])
        lower_red2 = np.array([160, 100, 100])
        upper_red2 = np.array([180, 255, 255])
        
        red_mask1 = cv2.inRange(hsv, lower_red1, upper_red1)
        red_mask2 = cv2.inRange(hsv, lower_red2, upper_red2)
        red_mask = cv2.bitwise_or(red_mask1, red_mask2)
        
        # Combine masks
        combined_mask = cv2.bitwise_or(white_mask, red_mask)
        
        # Apply morphological operations to reduce noise
        kernel = np.ones((3,3), np.uint8)
        combined_mask = cv2.morphologyEx(combined_mask, cv2.MORPH_CLOSE, kernel)
        combined_mask = cv2.morphologyEx(combined_mask, cv2.MORPH_OPEN, kernel)
        
        # Create debug visualization
        color_debug = np.dstack((combined_mask, combined_mask, combined_mask))
        color_debug[white_mask > 0] = [255, 255, 255]  # White
        color_debug[red_mask > 0] = [0, 0, 255]        # Red
        
        cv2.imshow('Color Threshold Debug', color_debug)
        
        return combined_mask
    def find_lane_pixels(self, binary_warped):
        """Find lane pixels using sliding window method with three lane detection"""
        # Take a histogram of the bottom half of the image
        histogram = np.sum(binary_warped[binary_warped.shape[0]//2:,:], axis=0)
        
        # Create debug visualization
        debug_img = np.dstack((binary_warped, binary_warped, binary_warped)) * 255
        
        # Plot histogram
        hist_img = np.zeros((100, binary_warped.shape[1], 3), dtype=np.uint8)
        histogram_scaled = (histogram * 100 / np.max(histogram)).astype(np.int32)
        for x, h in enumerate(histogram_scaled):
            cv2.line(hist_img, (x, 99), (x, 99-h), (0, 255, 0), 1)
        
        # Find peaks for three lanes
        # Divide image into three sections
        third_point = np.int32(histogram.shape[0]//3)
        
        # Find the maximum for each third of the image
        leftx_base = np.argmax(histogram[:third_point])
        middlex_base = np.argmax(histogram[third_point:2*third_point]) + third_point
        rightx_base = np.argmax(histogram[2*third_point:]) + 2*third_point
        
        # Draw peak points on histogram
        cv2.circle(hist_img, (leftx_base, 99-histogram_scaled[leftx_base]), 5, (255, 0, 0), -1)  # Blue
        cv2.circle(hist_img, (middlex_base, 99-histogram_scaled[middlex_base]), 5, (0, 255, 0), -1)  # Green
        cv2.circle(hist_img, (rightx_base, 99-histogram_scaled[rightx_base]), 5, (0, 0, 255), -1)  # Red
        
        cv2.imshow('Lane Detection Histogram', hist_img)
        
        # Set height of windows
        window_height = np.int32(binary_warped.shape[0]//self.nwindows)
        
        # Identify the x and y positions of all nonzero pixels in the image
        nonzero = binary_warped.nonzero()
        nonzeroy = np.array(nonzero[0])
        nonzerox = np.array(nonzero[1])
        
        # Current positions to be updated for each window
        leftx_current = leftx_base
        middlex_current = middlex_base
        rightx_current = rightx_base
        
        # Lists to store lane pixel indices
        left_lane_inds = []
        middle_lane_inds = []
        right_lane_inds = []
        
        # Step through the windows one by one
        for window in range(self.nwindows):
            win_y_low = binary_warped.shape[0] - (window + 1) * window_height
            win_y_high = binary_warped.shape[0] - window * window_height
            
            # Define window boundaries for all three lanes
            win_xleft_low = leftx_current - self.margin
            win_xleft_high = leftx_current + self.margin
            
            win_xmid_low = middlex_current - self.margin
            win_xmid_high = middlex_current + self.margin
            
            win_xright_low = rightx_current - self.margin
            win_xright_high = rightx_current + self.margin
            
            # Draw the windows on the debug image
            cv2.rectangle(debug_img, (win_xleft_low, win_y_low), 
                         (win_xleft_high, win_y_high), (255, 0, 0), 2)  # Blue
            cv2.rectangle(debug_img, (win_xmid_low, win_y_low), 
                         (win_xmid_high, win_y_high), (0, 255, 0), 2)  # Green
            cv2.rectangle(debug_img, (win_xright_low, win_y_low), 
                         (win_xright_high, win_y_high), (0, 0, 255), 2)  # Red
            
            # Identify the nonzero pixels in x and y within each window
            good_left_inds = ((nonzeroy >= win_y_low) & 
                            (nonzeroy < win_y_high) & 
                            (nonzerox >= win_xleft_low) & 
                            (nonzerox < win_xleft_high)).nonzero()[0]
            
            good_mid_inds = ((nonzeroy >= win_y_low) & 
                            (nonzeroy < win_y_high) & 
                            (nonzerox >= win_xmid_low) & 
                            (nonzerox < win_xmid_high)).nonzero()[0]
            
            good_right_inds = ((nonzeroy >= win_y_low) & 
                             (nonzeroy < win_y_high) & 
                             (nonzerox >= win_xright_low) & 
                             (nonzerox < win_xright_high)).nonzero()[0]
            
            # Append these indices to the lists
            left_lane_inds.append(good_left_inds)
            middle_lane_inds.append(good_mid_inds)
            right_lane_inds.append(good_right_inds)
            
            # If found enough pixels, recenter next window
            if len(good_left_inds) > self.minpix:
                leftx_current = np.int32(np.mean(nonzerox[good_left_inds]))
            if len(good_mid_inds) > self.minpix:
                middlex_current = np.int32(np.mean(nonzerox[good_mid_inds]))
            if len(good_right_inds) > self.minpix:
                rightx_current = np.int32(np.mean(nonzerox[good_right_inds]))
        
        # Concatenate the arrays of indices
        try:
            left_lane_inds = np.concatenate(left_lane_inds)
            middle_lane_inds = np.concatenate(middle_lane_inds)
            right_lane_inds = np.concatenate(right_lane_inds)
        except ValueError:
            return [], [], [], [], [], [], debug_img

        # Extract line pixel positions
        leftx = nonzerox[left_lane_inds]
        lefty = nonzeroy[left_lane_inds]
        middlex = nonzerox[middle_lane_inds]
        middley = nonzeroy[middle_lane_inds]
        rightx = nonzerox[right_lane_inds]
        righty = nonzeroy[right_lane_inds]

        # Color the detected pixels for debugging
        debug_img[lefty, leftx] = [255, 0, 0]    # Blue for left lane
        debug_img[middley, middlex] = [0, 255, 0]  # Green for middle lane
        debug_img[righty, rightx] = [0, 0, 255]    # Red for right lane
        
        return leftx, lefty, middlex, middley, rightx, righty, debug_img

    def fit_polynomial(self, leftx, lefty, middlex, middley, rightx, righty):
        """Fit polynomials to all three lanes"""
        # Minimum points required for fitting
        min_points = 10
        
        # Initialize fits
        left_fit = self.left_fit
        middle_fit = self.middle_fit
        right_fit = self.right_fit
        
        # Fit new polynomials if enough points are found
        if len(leftx) > min_points:
            left_fit = np.polyfit(lefty, leftx, 2)
        if len(middlex) > min_points:
            middle_fit = np.polyfit(middley, middlex, 2)
        if len(rightx) > min_points:
            right_fit = np.polyfit(righty, rightx, 2)
        
        # Apply smoothing if previous fits exist
        if self.left_fit is not None and left_fit is not None:
            left_fit = self.smooth_factor * self.left_fit + (1 - self.smooth_factor) * left_fit
        if self.middle_fit is not None and middle_fit is not None:
            middle_fit = self.smooth_factor * self.middle_fit + (1 - self.smooth_factor) * middle_fit
        if self.right_fit is not None and right_fit is not None:
            right_fit = self.smooth_factor * self.right_fit + (1 - self.smooth_factor) * right_fit
        
        # Store the new fits
        self.left_fit = left_fit
        self.middle_fit = middle_fit
        self.right_fit = right_fit
        
        return left_fit, middle_fit, right_fit

    def draw_lanes(self, binary_warped, left_fit, middle_fit, right_fit):
        """Draw all three lanes with distinct colors"""
        # Generate y values for plotting
        ploty = np.linspace(0, binary_warped.shape[0]-1, binary_warped.shape[0])
        
        # Calculate x values for each lane
        try:
            left_fitx = left_fit[0]*ploty**2 + left_fit[1]*ploty + left_fit[2]
            middle_fitx = middle_fit[0]*ploty**2 + middle_fit[1]*ploty + middle_fit[2]
            right_fitx = right_fit[0]*ploty**2 + right_fit[1]*ploty + right_fit[2]
        except (TypeError, AttributeError):
            return None
        
        # Create image to draw on
        out_img = np.dstack((binary_warped, binary_warped, binary_warped))*255
        
        # Draw left lane area (blue with transparency)
        if left_fit is not None and middle_fit is not None:
            pts_left = np.hstack((
                np.array([np.transpose(np.vstack([left_fitx, ploty]))]),
                np.array([np.flipud(np.transpose(np.vstack([middle_fitx, ploty])))])
            ))
            overlay = out_img.copy()
            cv2.fillPoly(overlay, np.int_([pts_left]), (255, 0, 0))
            out_img = cv2.addWeighted(overlay, 0.3, out_img, 0.7, 0)
        
        # Draw middle lane area (green with transparency)
        if middle_fit is not None and right_fit is not None:
            pts_middle = np.hstack((
                np.array([np.transpose(np.vstack([middle_fitx, ploty]))]),
                np.array([np.flipud(np.transpose(np.vstack([right_fitx, ploty])))])
            ))
            overlay = out_img.copy()
            cv2.fillPoly(overlay, np.int_([pts_middle]), (0, 255, 0))
            out_img = cv2.addWeighted(overlay, 0.3, out_img, 0.7, 0)
        
        # Draw the actual lane lines
        if left_fit is not None:
            for i in range(len(ploty)-1):
                cv2.line(out_img, 
                        (int(left_fitx[i]), int(ploty[i])), 
                        (int(left_fitx[i+1]), int(ploty[i+1])), 
                        (255, 200, 0), 4)  # Bright blue
        
        if middle_fit is not None:
            for i in range(len(ploty)-1):
                cv2.line(out_img, 
                        (int(middle_fitx[i]), int(ploty[i])), 
                        (int(middle_fitx[i+1]), int(ploty[i+1])), 
                        (0, 255, 0), 4)  # Green
        
        if right_fit is not None:
            for i in range(len(ploty)-1):
                cv2.line(out_img, 
                        (int(right_fitx[i]), int(ploty[i])), 
                        (int(right_fitx[i+1]), int(ploty[i+1])), 
                        (0, 200, 255), 4)  # Bright red
        
        # Add curve information
        font = cv2.FONT_HERSHEY_SIMPLEX
        if left_fit is not None:
            cv2.putText(out_img, f'Left curve: {left_fit[0]:.6f}', 
                       (30, 30), font, 0.7, (255, 255, 255), 2)
        if middle_fit is not None:
            cv2.putText(out_img, f'Middle curve: {middle_fit[0]:.6f}', 
                       (30, 60), font, 0.7, (255, 255, 255), 2)
        if right_fit is not None:
            cv2.putText(out_img, f'Right curve: {right_fit[0]:.6f}', 
                       (30, 90), font, 0.7, (255, 255, 255), 2)
        
        return out_img

    def image_callback(self, msg):
        try:
            # Convert ROS Image to OpenCV image
            cv_image = self.bridge.imgmsg_to_cv2(msg, "bgr8")
            
            # Extract ROI
            x, y, w, h = self.roi
            roi_image = cv_image[y:y+h, x:x+w]
            
            # Apply perspective transform
            warped = cv2.warpPerspective(roi_image, self.M, (w, h))
            
            # Apply color thresholding
            binary_warped = self.color_threshold(warped)
            
            # Find lane pixels for all three lanes
            leftx, lefty, middlex, middley, rightx, righty, window_img = self.find_lane_pixels(binary_warped)
            
            # Create debug image showing detected points
            debug_img = window_img.copy()
            
            # Color the detected points
            if len(leftx) > 0:
                debug_img[lefty, leftx] = [255, 0, 0]  # Blue for left lane
            if len(middlex) > 0:
                debug_img[middley, middlex] = [0, 255, 0]  # Green for middle lane
            if len(rightx) > 0:
                debug_img[righty, rightx] = [0, 0, 255]  # Red for right lane
            
            # Show debug visualization of detected points
            cv2.imshow('Detected Lane Points', debug_img)
            
            # Fit polynomials for all three lanes
            left_fit, middle_fit, right_fit = self.fit_polynomial(
                leftx, lefty, middlex, middley, rightx, righty)
            
            # Draw lanes with enhanced visualization
            if left_fit is not None or middle_fit is not None or right_fit is not None:
                lane_visualization = self.draw_lanes(binary_warped, left_fit, middle_fit, right_fit)
                
                if lane_visualization is not None:
                    # Unwarp the lane visualization
                    unwarped_lane = cv2.warpPerspective(lane_visualization, self.Minv, (w, h))
                    
                    # Combine with original image
                    result = cv_image.copy()
                    mask = (unwarped_lane != 0).any(axis=2)
                    result[y:y+h, x:x+w][mask] = cv2.addWeighted(
                        result[y:y+h, x:x+w][mask], 0.7,
                        unwarped_lane[mask], 0.3, 0)
                    
                    # Add debug information on the original image
                    font = cv2.FONT_HERSHEY_SIMPLEX
                    font_scale = 0.7
                    thickness = 2
                    color = (0, 255, 0)  # Green text
                    
                    # Display number of points detected for each lane
                    debug_info = [
                        f"Left Lane Points: {len(leftx)}",
                        f"Middle Lane Points: {len(middlex)}",
                        f"Right Lane Points: {len(rightx)}"
                    ]
                    
                    y_offset = 30
                    for i, text in enumerate(debug_info):
                        cv2.putText(result, text, 
                                  (30, y_offset + i*30),
                                  font, font_scale, color, 
                                  thickness)
                    
                    # Show all visualization windows
                    cv2.imshow('Original with Lanes', result)
                    cv2.imshow('Bird\'s Eye View with Lanes', lane_visualization)
                    cv2.imshow('Window View', window_img)
                    
                    # Show binary warped image for debugging
                    binary_debug = np.dstack((binary_warped, binary_warped, binary_warped)) * 255
                    cv2.imshow('Binary Warped', binary_debug)
                    
                    cv2.waitKey(1)
                    
                    # Publish processed image
                    processed_msg = self.bridge.cv2_to_imgmsg(result, "bgr8")
                    self.image_pub.publish(processed_msg)
                    
            else:
                rospy.logwarn("No lanes detected in this frame")
                
        except Exception as e:
            rospy.logerr(f"Error processing image: {str(e)}")
            rospy.logerr(f"Error occurred at line {e.__traceback__.tb_lineno}")

def main():
    try:
        detector = LaneDetector()
        
        # Print initialization message
        rospy.loginfo("Lane detector node is running. Press Ctrl+C to terminate.")
        
        # Keep the node running
        rospy.spin()
        
    except rospy.ROSInterruptException:
        rospy.loginfo("Lane detector node terminated.")
    except Exception as e:
        rospy.logerr(f"Unexpected error in lane detector node: {str(e)}")
    finally:
        # Cleanup
        cv2.destroyAllWindows()

if __name__ == '__main__':
    main()